File: src/main/java/com/genbridge/backend/services/QuestService.java
package com.genbridge.backend.services;

import com.genbridge.backend.dto.QuestCompletionRequest;
import com.genbridge.backend.dto.QuestRequest;
import com.genbridge.backend.entity.Quest;
import com.genbridge.backend.entity.QuestCompletion;

import java.util.List;
import java.util.Map;

/** Defines operations for managing quests and quest completions. */
public interface QuestService {
    List<Map<String, Object>> getPublishedQuestsForCurrentUser();
    Map<String, Object> getPublishedQuestByIdForCurrentUser(Long id);
    QuestCompletion completeQuest(Long questId, QuestCompletionRequest request);
    List<Map<String, Object>> getCurrentUserCompletions();
    List<Quest> getAllQuests();
    Quest createQuest(QuestRequest request);
    Quest updateQuest(Long id, QuestRequest request);
    void deleteQuest(Long id);
}

File: src/main/java/com/genbridge/frontend/src/services/questService.ts
import api from "./api";

export interface Quest {
  id: number;
  title: string;
  description: string;
  offlineInstruction: string;
  published: boolean;
  completed?: boolean;
  publishedAt?: string | null;
}

export interface QuestCompletion {
  id: number;
  questId: number;
  questTitle: string;
  reflection: string;
  completedAt: string;
}

export interface QuestRequest {
  title: string;
  description: string;
  offlineInstruction: string;
  published: boolean;
}

export interface QuestCompletionRequest {
  reflection: string;
}

const questService = {
  getQuests: async () => {
    const res = await api.get("/quests");
    return res.data;
  },

  getQuestById: async (id: string | number) => {
    const res = await api.get(`/quests/${id}`);
    return res.data;
  },

  completeQuest: async (id: string | number, payload: QuestCompletionRequest) => {
    const res = await api.post(`/quests/${id}/complete`, payload);
    return res.data;
  },

  getMyCompletions: async () => {
    const res = await api.get("/quests/completions");
    return res.data;
  },

  getAllAdminQuests: async () => {
    const res = await api.get("/admin/quests");
    return res.data;
  },

  createQuest: async (payload: QuestRequest) => {
    const res = await api.post("/quests", payload);
    return res.data;
  },

  updateQuest: async (id: string | number, payload: QuestRequest) => {
    const res = await api.put(`/quests/${id}`, payload);
    return res.data;
  },

  deleteQuest: async (id: string | number) => {
    const res = await api.delete(`/quests/${id}`);
    return res.data;
  },
};

export default questService;